code/modules/projectiles/projectile/experimental.dm

```diff
@@ -0,0 +1,644 @@
+	src.loc = newspawn
+
+	update_icon()
+
+/obj/item/projectile/ricochet/update_icon()//8 possible combinations
+	switch(pos_to)
+		if(NORTH)
+			if(pos_from == WEST)
+				dir = NORTHWEST
+			else
+				dir = EAST
+		if(SOUTH)
+			if(pos_from == WEST)
+				dir = WEST
+			else
+				dir = SOUTHEAST
+		if(EAST)
+			if(pos_from == NORTH)
+				dir = NORTHEAST
+			else
+				dir = SOUTH
+		if(WEST)
+			if(pos_from == NORTH)
+				dir = NORTH
+			else
+				dir = SOUTHWEST
+
+/obj/item/projectile/ricochet/Bump(atom/A as mob|obj|turf|area)
+	if(bumped)	return 0
+	bumped = 1
+
+	if(A)
+		if(istype(A,/turf/) || (istype(A,/obj/machinery/door/) && A.opacity))
+			var/obj/structure/ricochet_bump/bump = new(loc)
+			bump.dir = pos_to
+			playsound(get_turf(src), 'sound/items/metal_impact.ogg', 50, 1)
+			switch(pos_to)
+				if(NORTH)
+					if(pos_from == WEST)
+						pos_to = EAST
+					else
+						pos_to = WEST
+					pos_from = NORTH
+				if(SOUTH)
+					if(pos_from == WEST)
+						pos_to = EAST
+					else
+						pos_to = WEST
+					pos_from = SOUTH
+				if(EAST)
+					if(pos_from == NORTH)
+						pos_to = SOUTH
+					else
+						pos_to = NORTH
+					pos_from = EAST
+				if(WEST)
+					if(pos_from == NORTH)
+						pos_to = SOUTH
+					else
+						pos_to = NORTH
+					pos_from = WEST
+
+		else if(istype(A,/mob/living))//ricochet shots "never miss"
+			if(istype(A,/mob/living/carbon/human))
+				var/mob/living/carbon/human/H = A
+				if(istype(H.wear_suit,/obj/item/clothing/suit/armor/laserproof))// bwoing!!
+					var/obj/structure/ricochet_bump/bump = new(loc)
+					bump.dir = pos_to
+					playsound(get_turf(src), 'sound/items/metal_impact.ogg', 50, 1)
+					switch(pos_to)
+						if(NORTH)
+							if(pos_from == WEST)
+								pos_to = EAST
+							else
+								pos_to = WEST
+							pos_from = NORTH
+						if(SOUTH)
+							if(pos_from == WEST)
+								pos_to = EAST
+							else
+								pos_to = WEST
+							pos_from = SOUTH
+						if(EAST)
+							if(pos_from == NORTH)
+								pos_to = SOUTH
+							else
+								pos_to = NORTH
+							pos_from = EAST
+						if(WEST)
+							if(pos_from == NORTH)
+								pos_to = SOUTH
+							else
+								pos_to = NORTH
+							pos_from = WEST
+
+				else
+					A.bullet_act(src, def_zone)
+					spawn()
+						density = 0
+						invisibility = 101
+						//del(src)
+						loc = null
+						returnToPool(src)
+						OnDeath()
+			else
+				A.bullet_act(src, def_zone)
+				spawn()
+					density = 0
+					invisibility = 101
+					//del(src)
+					loc = null
+					returnToPool(src)
+					OnDeath()
+
+		else if(is_type_in_list(A,ricochet_bump))//beware fuel tanks!
+			A.bullet_act(src)
+			spawn()
+				density = 0
+				invisibility = 101
+				//del(src)
+				loc = null
+				returnToPool(src)
+				OnDeath()
+
+		else if((istype(A,/obj/structure/window) || istype(A,/obj/machinery/door/window)) && (A.loc == src.loc))//all this part is to prevent a bug that causes the shot to go through walls
```

Comment:
Then you need to check flags & ONBORDER, because this will happen for stuff like directional firelocks as well.